contracts/CallbackScheduler.cdc

```diff
@@ -0,0 +1,738 @@
+        }
+
+        // getNextID returns the next ID and increments the ID counter
+        access(self) fun getNextID(): UInt64 {
+            let nextID = self.nextID
+            self.nextID = self.nextID + 1
+            return nextID
+        }
+
+        // calculateScheduledTimestamp calculates the timestamp at which a callback 
+        // can be scheduled. It takes into account the priority of the callback and 
+        // the execution effort.
+        // - If the callback is low priority, it returns the lowPriorityScheduledTimestamp 
+        //    as a special value.
+        // - If the callback is high priority, it returns the timestamp if there is enough 
+        //    space or nil if there is no space left.
+        // - If the callback is medium priority and there is no space left it finds next 
+        //    available timestamp.
+        access(self) fun calculateScheduledTimestamp(
+            timestamp: UFix64, 
+            priority: Priority, 
+            executionEffort: UInt64
+        ): UFix64? {
+            if priority == Priority.Low {
+                return self.lowPriorityScheduledTimestamp
+            }
+
+            let used = self.slotUsedEffort[timestamp]
+            // if nothing is scheduled at this timestamp, we can schedule at provided timestamp
+            if used == nil { 
+                return timestamp
+            }
+            
+            let available = self.slotAvailableEffort(timestamp: timestamp, priority: priority)
+            // if theres enough space, we can schedule at provided timestamp
+            if executionEffort <= available {
+                return timestamp
+            } else if priority == Priority.High {
+                // high priority demands scheduling at exact timestamp or failing
+                return nil
+            }
+
+            // if there is no space left for medium priority we search for next available timestamp
+            // todo: check how big the callstack can grow and if we should avoid recursion
+            return self.calculateScheduledTimestamp(
+                timestamp: timestamp + 1.0, 
+                priority: priority, 
+                executionEffort: executionEffort
+            )
+        }
+
+        // slot available effort returns the amount of effort that is available for a given timestamp and priority.
+        access(self) fun slotAvailableEffort(timestamp: UFix64, priority: Priority): UInt64 {
+            let slotPriority: {CallbackScheduler.Priority: UInt64}? = self.slotUsedEffort[timestamp]
+            let priorityLimit = self.priorityEffortLimit[priority]!
+
+            if slotPriority == nil {
+                return priorityLimit
+            }
+
+            let highReserve = self.priorityEffortReserve[Priority.High]!
+            let mediumReserve = self.priorityEffortReserve[Priority.Medium]!
+
+            let highUsed = slotPriority![Priority.High] ?? 0
```

Comment:
Instead of having `slotPriority` as an optional, and force-unwrapping it everywhere, maybe just check for the priority and early return. Pseudo-code:

```cadence
if !self.slotUsedEffort.containsKey(timestamp) {
    return priorityLimit
}

let slotPriority = self.slotUsedEffort[timestamp]!
// ....
```